### PR TITLE
CXP-1359: Currency : Disable catalog if at least 1 FILTER becomes invalid

### PR DESCRIPTION
**Pull request by eoliveauakeneo**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: components/catalogs/back/src/Infrastructure/EventSubscriber/CurrencyDeactivationSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        $currency = $event->getSubject();
+
+        if (!$currency instanceof CurrencyInterface) {
+            return;
+        }
+
+        /** @var JobInstance $jobInstance */
+        $jobInstance = $this->jobInstanceRepository->findOneByIdentifier('disable_catalogs_on_currency_deactivation');
+
+        if ($jobInstance) {
```
[the remark this answers is not in the document]
Comment (eoliveauakeneo) [reply components/catalogs/back/src/Infrastructure/EventSubscriber/CurrencyDeactivationSubscriber.php:46]: @willy-ahva i removed it as we talked and the error come back when I do a make pim-test or pim-dev

image

File: components/catalogs/back/src/Infrastructure/Symfony/Resources/config/services.yml [hunk 2]
```diff
@@ -156,3 +156,49 @@ services:
...
+        public: false
+
+    akeneo_catalogs.job.disable_catalogs_on_currency_deactivation:
+        class: '%pim_connector.job.simple_job.class%'
+        arguments:
+            - 'disable_catalogs_on_currency_deactivation'
+            - '@event_dispatcher'
+            - '@akeneo_batch.job_repository'
+            -
+                - '@akeneo_catalogs.step.disable_catalogs_on_currency_deactivation'
```
[the remark this answers is not in the document]
Comment (eoliveauakeneo) [reply components/catalogs/back/src/Infrastructure/Symfony/Resources/config/services.yml:184]: i reused an existing configuration, i guess not at the moment

File: components/catalogs/back/tests/Integration/Infrastructure/Job/DisableCatalogsOnCurrencyDeactivationTaskletTest.php [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        ]);
+
+        $this->disableCurrency('EUR');
+        $this->waitForQueuedJobs();
+
+        $this->assertCatalogIsDisabled($idCatalogUS);
+        $this->assertCatalogIsEnabled($idCatalogFR);
+
+
+        $this->assertNotificationExistsForUsers([
```
Comment (RiffVentura) [inline components/catalogs/back/tests/Integration/Infrastructure/Job/DisableCatalogsOnCurrencyDeactivationTaskletTest.php:84]: Catalog deactivation knows nothing about notification: so your test should rather test if `InvalidCatalogDisabledEvent` is dispatched

File: components/catalogs/back/tests/Integration/Infrastructure/Persistence/Catalog/GetCatalogIdsContainingCurrenciesQueryTest.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    public function testItGetsCatalogsByCurrency(): void
+    {
+        $this->createUser('shopifi');
+        $catalogIdUS = 'db1079b6-f397-4a6a-bae4-8658e64ad47c';
+        $catalogIdFR = 'ed30425c-d9cf-468b-8bc7-fa346f41dd07';
+        $catalogIdUK = '27c53e59-ee6a-4215-a8f1-2fccbb67ba0d';
+
+        $this->createCatalog($catalogIdUS, 'Store US', 'shopifi');
```
Comment (RiffVentura) [inline components/catalogs/back/tests/Integration/Infrastructure/Persistence/Catalog/GetCatalogIdsContainingCurrenciesQueryTest.php:40]: `createCatalog` accepts a wider range of parameters to configure a catalog

File: components/catalogs/back/tests/Integration/Infrastructure/Persistence/Catalog/GetCatalogIdsContainingCurrenciesQueryTest.php [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        ]);
+
+        $resultBothCatalogs = $this->query->execute(['EUR']);
+        $this->assertEquals([$catalogIdUS, $catalogIdFR], $resultBothCatalogs);
+
+        $resultUSDCatalog = $this->query->execute(['USD']);
+        $this->assertEquals([$catalogIdUS], $resultUSDCatalog);
+
+        $resultNoCatalog = $this->query->execute(['GBP']);
+        $this->assertEquals([], $resultNoCatalog);
```
Comment (RiffVentura) [inline components/catalogs/back/tests/Integration/Infrastructure/Persistence/Catalog/GetCatalogIdsContainingCurrenciesQueryTest.php:62]: mixing test cases: IMO data providers should be used here instead

**Review by RiffVentura — APPROVED** [review]
:+1: 
Don't forget to update EE fixtures as well

File: components/catalogs/back/tests/Integration/Infrastructure/Job/DisableCatalogsOnCurrencyDeactivationTaskletTest.php (hunk 3 above)
Comment (eoliveauakeneo) [reply components/catalogs/back/tests/Integration/Infrastructure/Job/DisableCatalogsOnCurrencyDeactivationTaskletTest.php:84]: You're right i didn't though of checking the dispatch

File: components/catalogs/back/src/Application/Persistence/Catalog/GetCatalogIdsContainingCurrenciesQueryInterface.php [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+/**
+ * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface GetCatalogIdsContainingCurrenciesQueryInterface
+{
+    /**
+     * @param string[] $currencyCodes
+     * @return string[]
+     */
```
[the remark this answers is not in the document]
Comment (eoliveauakeneo) [reply components/catalogs/back/src/Application/Persistence/Catalog/GetCatalogIdsContainingCurrenciesQueryInterface.php:15]: we know they are arrays but not arrays of strings, I replaced it with array<string> as per willy's recommendations

File: components/catalogs/back/src/Infrastructure/EventSubscriber/CurrencyDeactivationSubscriber.php [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+    public function disableCatalogsIfCurrencyIsDisabled(GenericEvent $event): void
+    {
+        $currency = $event->getSubject();
+
+        if (!$currency instanceof CurrencyInterface || $currency->isActivated()) {
+            return;
+        }
+
+        /** @var JobInstance $jobInstance */
```
Comment (tseho) [inline components/catalogs/back/src/Infrastructure/EventSubscriber/CurrencyDeactivationSubscriber.php:44]:
```suggestion
        /** @var JobInstance|null $jobInstance */
```

**Review by tseho — APPROVED** [review]

**Review by masacc — APPROVED** [review]

File: components/catalogs/back/src/Infrastructure/Job/DisableCatalogsOnCurrencyDeactivationTasklet.php [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+use Akeneo\Catalogs\Application\Service\DispatchInvalidCatalogDisabledEventInterface;
+use Akeneo\Tool\Component\Batch\Model\StepExecution;
+use Akeneo\Tool\Component\Connector\Step\TaskletInterface;
+
+class DisableCatalogsOnCurrencyDeactivationTasklet implements TaskletInterface
+{
+    private ?StepExecution $stepExecution = null;
+
+    public function __construct(
+        private readonly GetCatalogIdsContainingCurrenciesQueryInterface $getCatalogsToDisableQuery,
```
[the remark this answers is not in the document]
Comment (eoliveauakeneo) [reply components/catalogs/back/src/Infrastructure/Job/DisableCatalogsOnCurrencyDeactivationTasklet.php:18]: Not really, updated

**Pull request merged**